internal/hashing/argon2id.go

```diff
@@ -0,0 +1,162 @@
+package hashing
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"golang.org/x/crypto/argon2"
+)
+
+const ARGON2_VERSION = 19
+
+type Argon2IDParameters struct {
+	SaltLength uint32
+	Time       uint32
+	Memory     uint32
+	Threads    uint8
+	KeyLength  uint32
+}
+
+// Default parameters recommended by OWASP
+var DefaultArgon2IDParameters = Argon2IDParameters{
+	SaltLength: 16,
+	Time:       2,
+	Memory:     19 * 1024, // 19 MiB
+	Threads:    1,
+	KeyLength:  32,
+}
+
+var ErrInvalidPassword = errors.New("invalid password")
+
+// Takes the password string, makes sure it's not empty, and hashes it using argon2id
+func HashPassword(password string) (string, error) {
+	if strings.TrimSpace(password) == "" {
+		return "", fmt.Errorf("%w: password cannot be empty", ErrInvalidPassword)
+	}
+	salt, err := generateSalt(DefaultArgon2IDParameters.SaltLength)
+	if err != nil {
+		return "", err
+	}
+	hash := hashPasswordWithSaltAndParams(password, salt, DefaultArgon2IDParameters)
+
+	encoded_salt := base64.RawStdEncoding.EncodeToString(salt)
+	encoded_hash := base64.RawStdEncoding.EncodeToString(hash)
+	hashedPassword := fmt.Sprintf("$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s",
+		ARGON2_VERSION,
+		DefaultArgon2IDParameters.Memory,
+		DefaultArgon2IDParameters.Time,
+		DefaultArgon2IDParameters.Threads,
+		encoded_salt,
+		encoded_hash,
+	)
+	return string(hashedPassword), nil
+}
+
+// CompareHashAndPassword takes a hashed password string and a password,
+// hashes the password with the same parameters as the hashed password and
+// compares the resulting value, returning an error if they do not match.
+//
+// When this function is passed an invalid hashed password string, it will
+// hash the password with the default parameters to prevent an attacker from
+// getting information from the timing of a login failure.
+func CompareHashAndPassword(hashedPassword string, password string) error {
```

Comment:
For reference, here are some upstream discussions on adding these helpers for argon2id:
- https://github.com/golang/go/issues/60740
- https://github.com/golang/go/issues/16971

Someone tried to add this feature to the upstream through the PR referenced below, but never got a review from the maintainers:
- https://go-review.googlesource.com/c/crypto/+/431595

I don't know if you looked at the implementation there, maybe we can use some of those ideas.

Reply:
I did not find those discussions when I did my researched, but found the third party library that is mentioned in the first link. Overall it looks similar, but I think the encoding/decoding idea is a good one and adding support for the `pepper` is also interesting.

Reply:
I took some inspiration from those, like using `crypto/subtle` and using `Strict()` for base64 decoding.